fixed conflict resolution behavior while apply podpresets

droot opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
This fixes the PodPreset application behavior in case of conflicts occur during the merging of Pod's information with PodPreset's. More details are in issue #47861 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes #47861

**Special notes for your reviewer**:
We are splitting the PodPreset application logic in two phases. In first phase, we try to detect the conflicts in information merging without modifying the Pod at all. If conflict occurs, then we reject the PodPresets injection. Incase of no conflicts, we apply the PodPresets and merge the information.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
PodPreset is not injected if conflict occurs while applying PodPresets to a Pod.
```

droot committed:
fixed conflict resolution behavior while apply podpresets

cblecker commented on hack/local-up-cluster.sh:
```diff
@@ -27,6 +27,8 @@ ALLOW_SECURITY_CONTEXT=${ALLOW_SECURITY_CONTEXT:-""}
 NODE_ADMISSION=${NODE_ADMISSION:-""}
 RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
+# RUNTIME_CONFIG=${RUNTIME_CONFIG:-"settingsio/v1alpha1=true,settings.k8s.io/v1alpha1/podpreset=true"}
+RUNTIME_CONFIG=${RUNTIME_CONFIG:-"api/all=true"}
```
What's your intended goal for "RUNTIME_CONFIG"? You're attempting to declare it on both lines 29 and 31.

droot replied:
This file got checked in by mistake. I have cleaned it up and enabled "api/all" by default.  PTAL.

cblecker replied:
@droot I don't see any changes to this file.

droot commented:
/assign @pmorie @jpeeler

droot commented:
/assign @jpeeler

pmorie commented on hack/local-up-cluster.sh:
```diff
@@ -26,7 +26,7 @@ ALLOW_PRIVILEGED=${ALLOW_PRIVILEGED:-""}
 PSP_ADMISSION=${PSP_ADMISSION:-""}
 NODE_ADMISSION=${NODE_ADMISSION:-""}
-RUNTIME_CONFIG=${RUNTIME_CONFIG:-""}
+RUNTIME_CONFIG=${RUNTIME_CONFIG:-"api/all=true"}
```
Are these debugging changes?

droot replied:
Took it out. It is not needed. I thought it was needed to enable "settings..../v1alpha1" API endpoint, but I was wrong, so took it out. PTAL.

cblecker commented:
/unassign

pmorie commented on hack/local-up-cluster.sh:
```diff
@@ -215,7 +215,7 @@ RKT_PATH=${RKT_PATH:-""}
 CHAOS_CHANCE=${CHAOS_CHANCE:-0.0}
 CPU_CFS_QUOTA=${CPU_CFS_QUOTA:-true}
-ENABLE_HOSTPATH_PROVISIONER=${ENABLE_HOSTPATH_PROVISIONER:-"false"}
+ENABLE_HOSTPATH_PROVISIONER=${ENABLE_HOSTPATH_PROVISIONER:-"true"}
```
This is also a debugging change, right? Def out of scpe for this PR

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
-		pod.ObjectMeta.Annotations[fmt.Sprintf("%s/podpreset-%s", annotationPrefix, pip.GetName())] = pip.GetResourceVersion()
+func safeToApplyPodPresetsOnContainer(ctr *api.Container, podPresets []*settings.PodPreset) error {
+	var errs []error
+	if _, err := MergeEnv(ctr.Env, podPresets); err != nil {
```
Seems like mergeEnv is probably unintentionally exported; we should make it private unless it's used elsewhere

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
-	// if there were no original envvar just return the pip envvar
-	if original == nil {
-		return pip.Spec.Env, nil
+func MergeEnv(envVars []api.EnvVar, podPresets []*settings.PodPreset) ([]api.EnvVar, error) {
```
Hrm, maybe it is intentional?

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
+
+	var mergedEnvFrom []api.EnvFromSource
+
+	mergedEnvFrom = append(mergedEnvFrom, envSources...)
```
Is there a reason you chose a different form of allocate and copy here vs mergeenv?

droot replied:
I followed the original code and didn't change it. I am trying to dig more in to why EnvFrom are treated differently. Quick look suggests, they should also be handled in the similar ways as Env, but I need to confirm that.

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
+
+			// make sure they are identical or throw an error
+			if !reflect.DeepEqual(found, v) {
+				errs = append(errs, fmt.Errorf("merging volume mounts for %s has a conflict on %s: \n%#v\ndoes not match\n%#v\n in container", pp.GetName(), v.Name, v, found))
```
It might be helpful if this identified the 'for' as a podpreset

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
+				continue
+			}
+
+			// make sure they are identical or throw an error
```
I'm wondering if identical but duplicated mounts is an error - use case?

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
-	for _, v := range original {
-		orig[v.MountPath] = v
-	}
+	for _, pp := range podPresets {
```
It seems like we can do all of this in a single nested loop, since this is orthogonal to the above loop but covers the same space

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
+				continue
+			}
+
+			// make sure they are identical or throw an error
```
Also not sure about duplicate but equal volumes.

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -334,3 +341,35 @@ func (c *podPresetPlugin) addEvent(pod *api.Pod, pip *settings.PodPreset, messag
+	}
+}
+
+func applyPodPresetsOnContainer(ctr *api.Container, podPresets []*settings.PodPreset) {
```
I wiuld add a comment explaining that it shiuld be safe to ignore these errors

droot reviewed (commented):
Thanks @pmorie for the review. Have addressed all comments except one related to handling merge of EnvFrom. Will be investigating more about it today. PTAL.

droot commented:
@pmorie addressed all the comments. Also updated the comment about EnvFrom. PTAL. (Also, I am not able to see review comment threads, thats very strange).

pmorie commented:
@k8s-bot ok to test

pmorie commented:
/release-note

droot commented:
/test pull-kubernetes-verify

droot commented:
/test pull-kubernetes-node-e2e

droot commented:
/assign @shiywang

pmorie commented on hack/local-up-cluster.sh:
```diff
@@ -406,8 +406,7 @@ function start_apiserver {
     # Admission Controllers to invoke prior to persisting objects in cluster
-    ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount${security_admission},ResourceQuota,DefaultStorageClass,DefaultTolerationSeconds
-
+    ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,PodPreset,ServiceAccount${security_admission},ResourceQuota,DefaultStorageClass,DefaultTolerationSeconds
```
If we were to do it again, I would say to make this change in another PR.

droot replied:
Makes sense. Took it out. Will make a separate PR for this.

pmorie commented on plugin/pkg/admission/podpreset/admission.go:
```diff
@@ -103,216 +104,217 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
+	glog.Infof("applied podpresets successfully on Pod: %s ", pod.GetGenerateName())
```
Is this always going to yield the right name?

droot replied:
Changed it to pod.Name. Also improved the logging a bit by making it more informative (added applied podpreset names as well)

droot replied:
pod.Name or pod.GetName() is empty, so kept pod.GenerateName(). I guess each pod instance will get a generated name.

pmorie reviewed (commented):
Just a couple nits; I also think godoc for each method would be fantastic to have.

droot commented:
@pmorie addressed the comments. Please take a look.

pmorie commented:
@k8s-bot test this

pmorie commented:
/test pull-kubernetes-kubemark-e2e-gce

pmorie commented:
/lgtm

pwittrock commented:
/approve

derekwaynecarr commented:
/approve

droot commented:
rebased with the master, so will need LGTM.

pmorie commented:
/lgtm
